src/as400_sha_fmt_plug.c

```diff
@@ -0,0 +1,419 @@
+// AS/400 SHA plugin for JtR
+// This software is Copyright (c) 2016 Bart Kulach (@bartholozz) and Rob Schoemaker (@5up3rUs3r)
+// and it is hereby released to the general public under the following terms:
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted.
+//
+// This plugin is loosely based on the lotus85 plugin by SÃ©bastien Kaczmarek <skaczmarek@quarkslab.com>
+//
+// AS/400 SHA1 hash is calculated as follows:
+// - userid is padded with spaces to be 10 characters long, 
+//   converted to uppercase and UTF-16BE
+// - password is converted to UTF-16BE
+// - Hash is calculated from SHA1(userid+password)
+// 
+// See http://www.hackthelegacy.org for details and tooling to retrieve hashes from AS/400 systems
+
+
+// ==================================================================================================
+// Registration of the plugin and includes
+
+#if FMT_EXTERNS_H
+extern struct fmt_main fmt_as400;
+#elif FMT_REGISTERS_H
+john_register_one(&fmt_as400);
+#else
+
+#include <stdio.h>
+#include <string.h>
+#include "stdint.h"
+
+#include "sha.h"
+
+#ifdef _OPENMP
+#include <omp.h>
+#define OMP_SCALE               64 
+static int omp_t = 1;
+#endif
+
+#include "formats.h"
+#include "common.h"
+#include "memdbg.h"
+
+
+// ==================================================================================================
+// Plugin definition
+
+#define FORMAT_LABEL          "as400-sha"
+#define FORMAT_NAME           "IBM i (aka AS/400)"
+#define ALGORITHM_NAME        "SHA1"
+#define BENCHMARK_COMMENT     ""
+#define BENCHMARK_LENGTH      -1
+#define PLAINTEXT_LENGTH      32
+#define CIPHERTEXT_LENGTH     40                                 // Length of ciphertext in hexadecimal text notation (like in inputfile)
+#define BINARY_SIZE           0
+#define BINARY_LENGTH         20                                 // Binary length of hashes
+#define BINARY_ALIGN          1
+#define SALT_SIZE             sizeof(struct custom_salt)
+#define SALT_ALIGN            4
+#define MIN_KEYS_PER_CRYPT    1
+#define MAX_KEYS_PER_CRYPT    1
+
+#define AS400_USERID_LENGTH   10                                 // Fixed length of AS/400 userid for calculations
+
+
+// ==================================================================================================
+// Structures and global variables
+
+// Charset used for testing if ciphertext string is hexadecimal
+static const char AS400_BASE16_CHARSET[] = "0123456789ABCDEFabcdef";
```

Comment:
Should we allow lower-case hex? If hashes "come" in uppercase there's no reason really.